File: exam_LRX/py/work7.py
import cv2
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import rcParams

# 设置中文字体
rcParams['font.sans-serif'] = ['SimHei']  # 使用黑体
rcParams['axes.unicode_minus'] = False   # 解决负号显示问题

def linear_transform(image, alpha=1.5, beta=50):
    """
    线性变换: g(x, y) = alpha * f(x, y) + beta
    :param image: 输入灰度图像
    :param alpha: 对比度增益
    :param beta: 亮度偏移
    :return: 线性变换后的图像
    """
    return cv2.convertScaleAbs(image, alpha=alpha, beta=beta)


def log_transform(image):
    """
    对数变换: g(x, y) = c * log(1 + f(x, y))
    :param image: 输入灰度图像
    :return: 对数变换后的图像
    """
    c = 255 / np.log(1 + np.max(image))
    log_image = c * np.log(1 + image.astype(np.float32))
    return np.uint8(log_image)


def gamma_transform(image, gamma=2.0):
    """
    伽马变换: g(x, y) = c * f(x, y)^gamma
    :param image: 输入灰度图像
    :param gamma: 伽马值
    :return: 伽马变换后的图像
    """
    c = 255.0 / (np.max(image) ** gamma)
    gamma_image = c * (image.astype(np.float32) ** gamma)
    return np.uint8(gamma_image)


def main():
    # 读取灰度图像
    image_path = '../image/image9.png'  # 替换为你的图片路径
    image = cv2.imread(image_path, cv2.IMREAD_GRAYSCALE)
    if image is None:
        print(f"无法读取图像文件：{image_path}")
        return

    # 线性变换
    linear_image = linear_transform(image, alpha=1.5, beta=50)

    # 对数变换
    log_image = log_transform(image)

    # 伽马变换
    gamma_image = gamma_transform(image, gamma=2.0)

    # 可视化结果
    plt.figure(figsize=(12, 8))
    plt.subplot(2, 2, 1), plt.title("原图片"), plt.imshow(image, cmap='gray'), plt.axis('off')
    plt.subplot(2, 2, 2), plt.title("线性变换"), plt.imshow(linear_image, cmap='gray'), plt.axis('off')
    plt.subplot(2, 2, 3), plt.title("对数变换"), plt.imshow(log_image, cmap='gray'), plt.axis('off')
    plt.subplot(2, 2, 4), plt.title("伽马变换"), plt.imshow(gamma_image, cmap='gray'), plt.axis('off')
    plt.tight_layout()
    plt.show()


if __name__ == "__main__":
    main()

File: exam_LRX/cpp/work7.cpp
//
// Created by liryi on 24-11-22.
// 7.给出图像增强处理中的灰度变换中的线性变换和非线性变换的实例。
#include <opencv2/opencv.hpp>
#include <iostream>
#include <cmath>

using namespace cv;
using namespace std;

// 线性变换
Mat linearTransform(const Mat& image, double alpha = 1.5, double beta = 50) {
    Mat result;
    image.convertTo(result, -1, alpha, beta);
    return result;
}

// 对数变换
Mat logTransform(const Mat& image) {
    Mat result, floatImage;
    image.convertTo(floatImage, CV_32F);
    float c = 255 / log(1 + 255); // 常数c
    log(floatImage + 1, result);
    result = result * c;
    result.convertTo(result, CV_8U);
    return result;
}

// 修正后的伽马变换
Mat gammaTransform(const Mat& image, double gamma = 2.0) {
    Mat result;
    unsigned char lookup[256];
    for (int i = 0; i < 256; i++) {
        lookup[i] = saturate_cast<uchar>(pow(i / 255.0, gamma) * 255.0);
    }
    LUT(image, Mat(1, 256, CV_8U, lookup), result);
    return result;
}

// 在图像上添加标识文字
void addLabel(Mat& image, const string& label) {
    putText(image, label, Point(10, 30), FONT_HERSHEY_SIMPLEX, 1, Scalar(255), 2);
}

int main() {
    // 读取灰度图像
    string imagePath = "E:\\DL_class\\exam_LRX\\cpp\\images\\image9.png";  // 替换为你的图片路径
    Mat image = imread(imagePath, IMREAD_GRAYSCALE);
    if (image.empty()) {
        cout << "无法加载图像: " << imagePath << endl;
        return -1;
    }

    // 线性变换
    Mat linearImage = linearTransform(image, 1.5, 50);

    // 对数变换
    Mat logImage = logTransform(image);

    // 伽马变换
    Mat gammaImage = gammaTransform(image, 2.0);

    // 添加标识
    addLabel(image, "Original image");
    addLabel(linearImage, "linear transformation");
    addLabel(logImage, "logarithm transformation");
    addLabel(gammaImage, "Gamma transformation");

    // 拼接图片
    Mat topRow, bottomRow, combined;
    hconcat(vector<Mat>{image, linearImage}, topRow);  // 将原图和线性变换拼接
    hconcat(vector<Mat>{logImage, gammaImage}, bottomRow);  // 拼接对数和伽马变换
    vconcat(topRow, bottomRow, combined);  // 上下拼接

    // 显示拼接后的结果
    namedWindow("Comparison of Results", WINDOW_NORMAL);
    imshow("Comparison of Results", combined);

    // 等待用户按下键退出
    waitKey(0);
    destroyAllWindows();
    return 0;
}

